internal/nri/plugin.go

```diff
@@ -0,0 +1,385 @@
+// existing CPU NUMA nodes and assigned GPU/MIG devices. It ensures that:
+// 1. Only valid system NUMA nodes are included from the container's existing nodes
+// 2. All NUMA nodes from assigned GPU/MIG devices are included
+func (p *NUMAFilterPlugin) filterContainerNUMANodes(c *api.Container) ([]string, error) {
+	if p.cache == nil {
+		return nil, fmt.Errorf("NUMA topology not initialized")
+	}
+
+	// Start with existing NUMA nodes from container, but only include valid system nodes
+	var numaNodes []int
+	if c.Linux != nil && c.Linux.Resources != nil && c.Linux.Resources.Cpu != nil {
+		mems := c.Linux.Resources.Cpu.Mems
+		// Parse NUMA nodes using k8s.io/utils/cpuset
+		nodes, err := cpuset.Parse(mems)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse NUMA nodes from mems %q: %w", mems, err)
+		}
+		// Only include nodes that exist in the system
+		for _, node := range nodes.List() {
+			if slices.Contains(p.cache.systemNodes, node) {
+				numaNodes = append(numaNodes, node)
+			}
+		}
+	}
+
+	// If no existing NUMA nodes were found, initialize with all system nodes
+	if len(numaNodes) == 0 {
+		numaNodes = append(numaNodes, p.cache.systemNodes...)
+	}
+
+	// Get assigned GPUs and MIG devices
+	uuids, err := p.getAssignedDevices(c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get assigned devices: %w", err)
+	}
+
+	// Add NUMA nodes for each UUID
+	for _, uuid := range uuids {
+		node, err := p.getDeviceNUMANode(uuid)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get NUMA node for device %s: %w", uuid, err)
+		}
+		numaNodes = append(numaNodes, node)
+	}
+
+	// Convert to strings
+	nodeStrs := make([]string, len(numaNodes))
+	for i, node := range numaNodes {
+		nodeStrs[i] = strconv.Itoa(node)
+	}
+
+	return nodeStrs, nil
+}
+
+// getAssignedDevices returns the list of GPU and MIG device UUIDs assigned to a container.
+// The function should branch based on the configuration in p.devicePluginConfig to determine how to
+// handle different device types (e.g., MIG devices, replicated resources) and their
+// NUMA node assignments.
+func (p *NUMAFilterPlugin) getAssignedDevices(c *api.Container) ([]string, error) {
+	// TODO: Implement branching logic based on p.devicePluginConfig to handle
+	// different device list strategies
+
+	// Look for NVIDIA_VISIBLE_DEVICES environment variable
+	for _, env := range c.Env {
+		if strings.HasPrefix(env, "NVIDIA_VISIBLE_DEVICES=") {
+			// Get the value after the equals sign
+			value := strings.TrimPrefix(env, "NVIDIA_VISIBLE_DEVICES=")
+
+			// Handle special cases
+			switch value {
+			case "", "void":
+				return nil, nil
+			case "all":
+				uuids, err := p.getAllDeviceUUIDs()
+				if err != nil {
+					return nil, fmt.Errorf("failed to get all device UUIDs: %w", err)
+				}
+				return uuids, nil
+			default:
+				// Handle comma-separated list of UUIDs
+				return strings.Split(value, ","), nil
```

Comment:
Can `value` at times be a list of integers instead of a list of UUIDs?

Reply:
Yes, that is why I have the todo at the top of this saying we need to handel all `list-strategies`. I didn't include it in the list in the PR description though -- adding now.